Automattic/wp-calypso
# ETK: Improve webpack build

This is one of my first attempts at digging into webpack configs, so I'd love feedback about how to improve this approach. What alternatives might make more sense?

### Changes proposed in this Pull Request
Updates the webpack build in the editing toolkit plugin in a few ways:
1. Removes the need for the odd package.json entries for each module.
2. Bundles all assets into a root "dist/" directory, rather than having a separate "dist/" directory for each plugin.
3. Creates a function called `use_webpack_assets`, which allows us to access webpack bundled assets in a standardized manner -- and helps remove several hundred LOC.
4. Updates all modules to load with this new format.

#### Webpack changes

The webpack config changes work like this:
1. Parses the files in the `editing-toolkit-plugin` directory to find each subdirectory which contains an `index.(j|t)s` file.
2. Removes any entries which are part of the `excludedSources` list in package.json
3. Adds other sources from `additionalSources`
4. Standardizes everything such that all sources are compiled into the top-level of the "dist" directory.

#### Asset enqueue changes
I'd been wanting to write something like this for a while, because I've noticed that projects setting up in the plugin tend to do things slightly differently with this part of the code. It is about 20 LOC which needs to be repeated for every project for setup, so this improves things a lot:

1. Less code to maintain
2. You only have to remember one function to use when setting up new code.
3. All code is registered similarly, so things like translations work "out of the box".

#### Excluding unminified JS code from the final build
>  Building into a single dist folder and syncing only that would make for smaller and less bloated deploys

@jsnajdr this might be easier to accomplish now that all JS is built to the same directory, but there are still some challenges here:
1. Not all JS in the plugin is compiled. For example, when loading front-end assets, some modules just point at an es5 file locally without having it compiled by webpack.
2. There is PHP code.

My first thought was that we could simply remove all .js code from the directory as part of a deploy script. But then we'd have to keep the non-compiled versions.

### In progress:
Newspack block assets are 404ing, so that needs to be debugged and fixed.

### Changing public API

The public API for the build commands will change: it will not be possible to run `--sync` for the dev or build commands any longer. Instead, you should run the wpcom-sync command directly from a different terminal, or in the same terminal like so: `yarn dev & yarn wpcom-sync`. We could add this back, but it makes things a bit easier to maintain, and I don't think there's a huge cost:

1. We don't need to maintain an extra script which is just a proxy to a different script.
2. Folks might learn how to use the scripts a bit better if they have to use the lower-level version without the magic.
3. It's really just a small syntax change, so the only impact will be folks having to re-learn this command.
4. It is more consistent with the other apps using this sync script, like `wpcom-block-editor`.

### Testing instructions
1. Run `yarn build` or `yarn dev`
2. Compiled assets should work properly.
3. All CI tests should pass.

Related to #48833
